File: nodejs/src/registry.ts
/** Framework-neutral tool registry for MCP, REST adapters, and OpenAPI adapters. */
import { createState, handleRequest, type JsonRpcRequest, type JsonRpcResponse } from './dispatch.js'
import { toJsonSchema, type InputSchema } from './schema.js'
import { buildOpenApiSpec, type OpenApiToolSource } from './openapi.js'

export type RouteMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE'
export type McpHandler = (request: JsonRpcRequest) => Promise<JsonRpcResponse | null>

export interface Tool<TEnv = Record<string, unknown>> {
  description: string
  input: InputSchema
  route?: { method: RouteMethod; path: string }
  execute: (args: Record<string, unknown>, env: TEnv) => Promise<unknown>
}

export interface RouteDefinition<TEnv = Record<string, unknown>> {
  name: string
  method: RouteMethod
  path: string
  tool: Tool<TEnv>
}

export interface RegistryOptions<TEnv> {
  getEnv?: () => TEnv
  /** Server name/version reported in `initialize` and the OpenAPI `info` block. Defaults to 'ZeroMCP' / '1.0.0'. */
  title?: string
  version?: string
  /** Default per-tool execute timeout in ms, unless a tool overrides it. Defaults to 30_000. */
  executeTimeout?: number
}

export interface ToolRegistry<TEnv = Record<string, unknown>> {
  routes: RouteDefinition<TEnv>[]
  openapi: Record<string, unknown>
  mcp: McpHandler
}

export function createRegistry<TEnv = Record<string, unknown>>(
  tools: Record<string, Tool<TEnv>>,
  options: RegistryOptions<TEnv> = {},
): ToolRegistry<TEnv> {
  const getEnv = options.getEnv ?? (() => ({} as TEnv))
  const toolMap = new Map<string, OpenApiToolSource & { input: InputSchema; execute: (args: Record<string, unknown>) => Promise<unknown> }>()

  for (const [name, tool] of Object.entries(tools)) {
    toolMap.set(name, {
      description: tool.description,
      input: tool.input,
      cachedSchema: toJsonSchema(tool.input),
      route: tool.route,
      execute: (args: Record<string, unknown>) => tool.execute(args, getEnv()),
    })
  }

  const routeCounts = new Map<string, number>()
  const routes: RouteDefinition<TEnv>[] = []
  for (const [name, tool] of Object.entries(tools)) {
    if (!tool.route) continue
    const key = `${tool.route.method} ${tool.route.path}`
    routeCounts.set(key, (routeCounts.get(key) ?? 0) + 1)
    routes.push({ name, method: tool.route.method, path: tool.route.path, tool })
  }
  for (const [key, count] of routeCounts) {
    if (count > 1) console.error(`[zeromcp] ${count} tools registered the same route "${key}" — only your own router's route matching, not this registry, decides which one runs`)
  }

  const state = createState({
    tools: toolMap,
    executeTimeout: options.executeTimeout ?? 30_000,
    version: options.version ?? '1.0.0',
  })

  return {
    routes,
    openapi: buildOpenApiSpec(toolMap, { title: options.title, version: options.version }),
    mcp: (request: JsonRpcRequest) => handleRequest(request, state),
  }
}

File: swift/Sources/ZeroMcp/Registry.swift
/// One tool exposed as an HTTP-style route.
/// Named `RegistryRoute` (not `RouteDefinition`) because `RouteDefinition`
/// is already taken — it's the existing `{ method, path }` struct attached
/// to a tool (Tool.swift:46). This type is the registry's richer route
/// *entry*: the route metadata plus the tool it belongs to.
public struct RegistryRoute {
    public let name: String
    public let method: String   // uppercased, e.g. "GET" — matches RouteDefinition.method
    public let path: String     // ":param" segments, unmatched — matches RouteDefinition.path
    public let tool: ToolDefinition
}

/// Signature of the registry's JSON-RPC handler. Matches
/// `ZeroMcp.handleRequest(_:)` exactly (untyped dictionaries — this
/// codebase has no JsonRpcRequest/JsonRpcResponse types to reuse or port).
public typealias McpHandler = (_ request: [String: Any]) async -> [String: Any]?

public struct RegistryOptions {
    /// Config used to build the internal dispatch instance (title for
    /// OpenAPI info, executeTimeout default, etc). Defaults to an empty
    /// ZeroMcpConfig() — NOT ZeroMcpConfig.load(), which reads
    /// ./zeromcp.config.json from the current working directory as a
    /// side effect. An embeddable library must not do ambient
    /// filesystem I/O by default; pass ZeroMcpConfig.load() explicitly
    /// if that behavior is wanted.
    public var config: ZeroMcpConfig

    public init(config: ZeroMcpConfig = ZeroMcpConfig()) {
        self.config = config
    }
}

public struct Registry {
    public let routes: [RegistryRoute]
    public let openapi: [String: Any]
    public let mcp: McpHandler
}

/// Build a framework-neutral registry from a plain tool map.
///
/// `tools` uses the same `ToolDefinition` struct the `.tool()` builder on
/// `ZeroMcp` produces internally (Tool.swift:56) — construct it directly
/// via `ToolDefinition(description:input:permissions:route:execute:)`
/// without going through a `ZeroMcp` instance's builder DSL.
public func createRegistry(
    _ tools: [String: ToolDefinition],
    options: RegistryOptions = RegistryOptions()
) -> Registry {
    let instance = ZeroMcp(config: options.config)
    instance.tools = tools

    let routes = tools
        .filter { $0.value.route != nil }
        .sorted(by: { $0.key < $1.key })
        .map { name, tool -> RegistryRoute in
            let route = tool.route!
            return RegistryRoute(name: name, method: route.method, path: route.path, tool: tool)
        }

    return Registry(
        routes: routes,
        openapi: instance.buildOpenApiSpec(),
        mcp: { request in await instance.handleRequest(request) }
    )
}
